File: LeChangeNewDemo/LeChangeDemo/Business/DeviceList/View/LCDeviceListViewController.m
//
//  Copyright © 2019 dahua. All rights reserved.
//

#import "LCDeviceListViewController.h"
#import "LCDeviceListPresenter.h"
#import <KVOController/KVOController.h>
#import <MJRefresh/MJRefresh.h>


@interface LCDeviceListViewController ()

/// presenter
@property (strong, nonatomic) LCDeviceListPresenter *presenter;


@end

@implementation LCDeviceListViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    [self.presenter initSDKLog];
    [self.presenter initSDK];
    [self setupDeviceListView];
    [self addObserver];
    [LCProgressHUD showHudOnView:nil];
    
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    weakSelf(self);
    [self.deviceListView.mj_header beginRefreshing];
    [self lcCreatNavigationBarWith:LCNAVIGATION_STYLE_DEVICELIST buttonClickBlock:^(NSInteger index) {
        if (index == 1) {
            [weakself.navigationController pushToAddDeviceScanPage];
        } else {
            [weakself.navigationController popViewControllerAnimated:YES];
        }
    }];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
}

- (LCDeviceListPresenter *)presenter {
    if (!_presenter) {
        _presenter = [LCDeviceListPresenter new];
        _presenter.listContainer = self;
    }
    return _presenter;
}

- (void)addObserver {
    weakSelf(self);
    [self.KVOController observe:self.presenter keyPath:@"lcDevices" options:NSKeyValueObservingOptionNew block:^(id  _Nullable observer, id  _Nonnull object, NSDictionary<NSString *,id> * _Nonnull change) {
        [weakself.deviceListView reloadData];
    }];
    [self.KVOController observe:self.presenter keyPath:@"openDevices" options:NSKeyValueObservingOptionNew block:^(id  _Nullable observer, id  _Nonnull object, NSDictionary<NSString *,id> * _Nonnull change) {
        [weakself.deviceListView reloadData];
    }];
}

- (void)removeObserver {
    [self.KVOController unobserve:self.presenter];
}

- (void)setupDeviceListView {
    
    self.deviceListView = [UITableView new];
    [self.view addSubview:self.deviceListView];
    self.deviceListView.separatorStyle = UITableViewCellSeparatorStyleNone;
    weakSelf(self);
    self.deviceListView.dataSource = self.presenter;
    self.deviceListView.delegate = self.presenter;
    [self.deviceListView registerClass:NSClassFromString(@"LCDeviceListCell") forCellReuseIdentifier:@"LCDeviceListCell"];
    [self.deviceListView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.mas_equalTo(self.view.mas_top);
        make.left.right.mas_equalTo(weakself.view);
        make.bottom.mas_equalTo(weakself.view.mas_bottom).offset(-kBottomSafeHeight);
    }];
    self.deviceListView.rowHeight = UITableViewAutomaticDimension;
	self.deviceListView.estimatedRowHeight = 0;
    [self.deviceListView lc_setEmyptImageName:@"common_pic_nodevice" andDescription:@"device_manager_list_no_device".lc_T];
    self.deviceListView.mj_header = [LCRefreshHeader headerWithRefreshingTarget:self.presenter refreshingAction:@selector(refreshData:)];
    self.deviceListView.mj_footer = [LCRefreshFooter footerWithRefreshingTarget:self.presenter refreshingAction:@selector(loadMoreData:)];
}

- (void)dealloc {
    [self removeObserver];
}
@end

File: LeChangeNewDemo/LeChangeDemo/Business/DeviceList/View/LCDeviceListViewController.swift
//
//  LCDeviceListViewController.swift
//  LeChangeDemo
//
//  Created by yyg on 2022/9/24.
//  Copyright © 2022 Imou. All rights reserved.
//

import SnapKit
import LCBaseModule
import KVOController
import UIKit

@objcMembers class LCDeviceListViewController : LCBasicViewController {
    /// presenter
    lazy var presenter: LCDeviceListPresenter = {
        let p = LCDeviceListPresenter()
        p.listContainer = self
        return p
    }()
    /// 设备列表视图
    lazy var deviceListView: UITableView = {
        let tableView = UITableView()
        tableView.separatorStyle = .none
        tableView.dataSource = self.presenter
        tableView.delegate = self.presenter
        tableView.register(LCDeviceListCell.self, forCellReuseIdentifier: "LCDeviceListCell")
        tableView.rowHeight = UITableViewAutomaticDimension
        tableView.estimatedRowHeight = 0
        tableView.backgroundColor = .clear
        tableView.tableFooterView = UIView(frame: CGRect(x: 0, y: 0, width: self.view.frame.size.width, height: 15))
        tableView.lc_setEmyptImageName("common_pic_nodevice", andDescription: "device_manager_list_no_device".lc_T)
        let header = LCRefreshHeader.init {[weak self] in
            self?.presenter.refreshData()
        }
        tableView.mj_header = LCRefreshHeader.init {[weak self] in
            self?.presenter.refreshData()
        }
        tableView.mj_footer = LCRefreshFooter.init(refreshingBlock: { [weak self] in
            self?.presenter.loadMoreData()
        })
        
        return tableView
    }()

    override func viewDidLoad() {
        super.viewDidLoad()
        self.presenter.initSDKLog()
        self.presenter.initSDK()
        self.setupDeviceListView()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.deviceListView.mj_header?.beginRefreshing()
        self.navigationController?.isNavigationBarHidden = false
        self.navigationController?.navigationBar.setBarBackgroundColor(color: .white)
        self.lcCreatNavigationBar(with: LCNAVIGATION_STYLE_DEVICELIST) {[weak self] index in
            if index == 1 {
                self?.showGuideAlert {
                    self?.navigationController?.pushToAddDeviceScanPage()
                }
            } else {
                self?.navigationController?.popViewController(animated: true)
            }
        }
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
    }
    
    func setupDeviceListView() {
        self.view.addSubview(self.deviceListView)
        self.deviceListView.snp.makeConstraints({ make in
            make.top.leading.trailing.equalToSuperview()
            make.bottom.equalToSuperview().offset(LC_bottomSafeMargin)
        })
    }

    deinit {
        
    }
    
    func showGuideAlert(action: @escaping () -> ()) {
        LCAlertView.lc_ShowAlert(title: "Alert_Title_Notice2".lc_T, detail: "open_local_network_to_add_device".lc_T, confirmString: "common_set".lc_T, cancelString: "common_got_it".lc_T) { isConfirmSelected in
            if isConfirmSelected == true {
                self.goToConfigLocalNetWork()
            } else {
                action()
            }
        }
    }
    
    /// 跳转设置wifi
    private func goToConfigLocalNetWork() {
        guard let url = URL.init(string: UIApplicationOpenSettingsURLString) else {
            return
        }
        if #available(iOS 10.0, *) {
            //先判断是否有iOS10SDK的方法，如果有，则实现iOS10的跳转
            if UIApplication.shared.canOpenURL(url) {
                UIApplication.shared.open(url, options: [:], completionHandler: nil)
            }
        } else {
            // Fallback on earlier versions
            if UIApplication.shared.canOpenURL(url) {
                UIApplication.shared.openURL(url)
            }
        }
    }
    
}
